### ISSUE TITLE
[sdc] Add. Sense: Logical unit communication failure

### ISSUE THREAD

**Issue by zjs1224522500** [open-iscsi/tcmu-runner#628]
Hi guys:
- I wrote a tcmu_handler refer to `file_optical` handler. And I build and install it. Then I create iscsi target and block device with `targetcli`, everything is ok. 
- At last, in client, i use isciadm to establish session with block device backed by my own stoarge implementation. The command `fdisk -l` show the right block device info. I tried to use `mkfs.ext4` to format this block device. And it took long time to response with error message such as

```
[root@45 ~]# mkfs.ext4 -vv /dev/sdb
mke2fs 1.42.9 (28-Dec-2013)
/dev/sdb is entire device, not just one partition!
Proceed anyway? (y,n) y
fs_types for mke2fs.conf resolution: 'ext4'
Warning: could not erase sector 2: Attempt to write block to filesystem resulted in short write
Filesystem label=
OS type: Linux
Block size=4096 (log=2)
Fragment size=4096 (log=2)
Stride=0 blocks, Stripe width=16 blocks
65536 inodes, 262144 blocks
13107 blocks (5.00%) reserved for the super user
First data block=0
Maximum filesystem blocks=268435456
8 block groups
32768 blocks per group, 32768 fragments per group
8192 inodes per group
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376

Allocating group tables: done                            
Warning: could not read block 0: Attempt to read block from filesystem resulted in short read
Warning: could not erase sector 0: Attempt to write block to filesystem resulted in short write
Writing inode tables: done                            
Creating journal (8192 blocks): done
Writing superblocks and filesystem accounting information: 0/8
Warning, had trouble writing out superblocks.
```
- However, in iscis server end, the I/O operation is running correctly. And I logged the I/O operations in Server end with tcmu_log so I am sure the **iscsi server** is running correctly.
- And I have no clues to find out the problems. And I tried to use `dmesg` in **iscsi client end**, there are a lot of error msgs such as 
```
[2020/06/04 14:16:38] [23219.279774] Buffer I/O error on dev sdc, logical block 0, async page read
[2020/06/04 14:17:09] [23250.254420] sd 10:0:0:0: [sdc] FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[2020/06/04 14:17:09] [23250.254438] sd 10:0:0:0: [sdc] Sense Key : Not Ready [current] 
[2020/06/04 14:17:09] [23250.254445] sd 10:0:0:0: [sdc] Add. Sense: Logical unit communication failure
[2020/06/04 14:17:09] [23250.254453] sd 10:0:0:0: [sdc] CDB: Read(10) 28 00 00 00 00 00 00 00 08 00
[2020/06/04 14:17:09] [23250.254459] blk_update_request: I/O error, dev sdc, sector 0
[2020/06/04 14:17:09] [23250.254466] Buffer I/O error on dev sdc, logical block 0, async page read
[2020/06/04 14:17:09] [23250.254520] Dev sdc: unable to read RDB block 0
[2020/06/04 14:17:40] [23281.357078] sd 10:0:0:0: [sdc] FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[2020/06/04 14:17:40] [23281.357087] sd 10:0:0:0: [sdc] Sense Key : Not Ready [current] 
[2020/06/04 14:17:40] [23281.357093] sd 10:0:0:0: [sdc] Add. Sense: Logical unit communication failure
[2020/06/04 14:17:40] [23281.357099] sd 10:0:0:0: [sdc] CDB: Read(10) 28 00 00 00 00 00 00 00 08 00
[2020/06/04 14:17:40] [23281.357103] blk_update_request: I/O error, dev sdc, sector 0
[2020/06/04 14:17:40] [23281.357108] Buffer I/O error on dev sdc, logical block 0, async page read
[2020/06/04 14:18:11] [23312.331748] sd 10:0:0:0: [sdc] FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[2020/06/04 14:18:11] [23312.331766] sd 10:0:0:0: [sdc] Sense Key : Not Ready [current] 
[2020/06/04 14:18:11] [23312.331774] sd 10:0:0:0: [sdc] Add. Sense: Logical unit communication failure
[2020/06/04 14:18:11] [23312.331782] sd 10:0:0:0: [sdc] CDB: Read(10) 28 00 00 00 00 00 00 00 08 00
[2020/06/04 14:18:11] [23312.331788] blk_update_request: I/O error, dev sdc, sector 0
[2020/06/04 14:18:11] [23312.331795] Buffer I/O error on dev sdc, logical block 0, async page read
[2020/06/04 14:18:11] [23312.331845]  sdc: unable to read partition table
[2020/06/04 14:18:42] [23343.306412] sd 10:0:0:0: [sdc] FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[2020/06/04 14:18:42] [23343.306435] sd 10:0:0:0: [sdc] Sense Key : Not Ready [current] 
[2020/06/04 14:18:42] [23343.306443] sd 10:0:0:0: [sdc] Add. Sense: Logical unit communication failure
[2020/06/04 14:18:42] [23343.306453] sd 10:0:0:0: [sdc] CDB: Read(10) 28 00 00 1f ff 80 00 00 08 00
[2020/06/04 14:18:42] [23343.306459] blk_update_request: I/O error, dev sdc, sector 2097024
[2020/06/04 14:19:13] [23374.281196] sd 10:0:0:0: [sdc] FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[2020/06/04 14:19:13] [23374.281218] sd 10:0:0:0: [sdc] Sense Key : Not Ready [current] 
[2020/06/04 14:19:13] [23374.281226] sd 10:0:0:0: [sdc] Add. Sense: Logical unit communication failure
[2020/06/04 14:19:13] [23374.281235] sd 10:0:0:0: [sdc] CDB: Read(10) 28 00 00 1f ff 80 00 00 08 00
[2020/06/04 14:19:13] [23374.281242] blk_update_request: I/O error, dev sdc, sector 2097024
[2020/06/04 14:19:13] [23374.281251] Buffer I/O error on dev sdc, logical block 262128, async page read
```
- So I am not sure about if my own backend storage implementation exists some errors 
or
- there exists some errors when communicate with iscsi.
---
Could you give any clues to find out problems? If need more info, please feel free to tell me.
BTW, is there any test program to test the storage backend implementation? If there is, it will be helpful!

Thx & BR

**Comment by lxbsz** [comment]
It should be your handler's issue, the read should be not successful.

Could you enable the highest debug mode in tcmu-runner and try again ? Please attach the tcmu-runner.log.

**Comment by zjs1224522500** [comment]
@lxbsz Thanks for your reply in advance.
Because the log is too long, I used the [github gist:tcmu-runner.log](https://gist.github.com/zjs1224522500/3f93fbac4b3b2e04f0e321185c5b2f18)  to show the log.

**Comment by lxbsz** [comment]
Could you set 'tcmur_cmd_time_out=10' when creating the target device and try to catch the log again ? 

With this we can see some useful logs if your handler couldn't response the cmd requests correctly.

**Comment by zjs1224522500** [comment]
- I tried to add `tcmur_cmd_time_out=20` when createing the target device. And log is in [gist](https://gist.github.com/zjs1224522500/663b69a15557ef49047d55299274e654).
- I have to say, when the `mkfs.ext4` failed, the target device is still working. So previous `tcmu-runner.log` may lack some log. In this [log](https://gist.github.com/zjs1224522500/663b69a15557ef49047d55299274e654), I waited the operations for a period. Beccause in my test sceneriao, all operations will takel a long time to complete, so in log file only shows the partial log. And I will upload the whole log when all operations finish.
- Although I didn't find the big differences compared with previous log.
